zebra-rpc/src/methods.rs

```diff
@@ -902,15 +901,23 @@ where
                 })?;
 
             match response {
-                zebra_state::ReadResponse::Transaction(Some((tx, height))) => Ok(
-                    GetRawTransaction::from_transaction(tx, Some(height), verbose != 0).map_err(
-                        |error| Error {
-                            code: ErrorCode::ServerError(0),
-                            message: error.to_string(),
-                            data: None,
-                        },
-                    )?,
-                ),
+                zebra_state::ReadResponse::Transaction(Some((tx, height))) => {
+                    // # Correctness:
+                    //
+                    // The transaction request only checks the best chain, so it's okay
+                    // to get the transaction depth by subtracting the transaction height
+                    // from the best chain tip height just before the transaction request.
+                    let depth = best_tip_height
+                        .map(|tip_height| (tip_height.0 - height.0))
```

Comment:
If the best chain changes, and the old `tip_height` is below the transaction height in the latest best chain, this subtraction will:
* panic in debug mode
* produce a large invalid value in release mode

Instead, we could:
* return the depth in the state request, so it is always valid and consistent
* replace the result with `0` (but that might cause problems for clients, because that's incorrect data - we know the transaction is now in the best chain)

I'd like to avoid re-checking the best tip, because the same chain-switch issue can happen again between the state request and the re-check.

Reply:
We should probably also add a test with this edge-case, so we don't make this mistake in future RPC changes. (We almost missed it and merged the code.)

Reply:
Added a test in [`ef89d82`](https://github.com/ZcashFoundation/zebra/pull/6287/commits/ef89d82dc013db547bab3ee5987926359df6b97d), and updated the method to get confirmations from the state service in [`2610124`](https://github.com/ZcashFoundation/zebra/pull/6287/commits/26101243da0af4652dc281448084d51249a23a31).